### DISCUSSION TITLE
Bounding Volume Hierarchy

### DISCUSSION THREAD

**Discussion by kurt-rhee** [Enigmatisms/AdaPT#18] [Q&A] [answered]
Hey, 

This project is really amazing, I was interested in the BVH that you implemented.  Is it something that you wrote or is it something that you were able to pull from another source?

Thank you!

Kurt

**Comment by Enigmatisms** [comment]
Hi Kurt
I'm not quite sure about what 'pulling from another source' means here. So I can only reply you based on my own understanding.

This BVH is something I wrote myself when I was learning the PBR-book. Actually PBR book provides an excellent implementation, but I thought if I just copied the code then there wouldn't be much I could learn. This code is based on the understanding of mine for the contents in the PBR-book, therefore the idea behind the BVH implementation should closely resemble to that of PBR-book's.

And, maybe, if I understand correctly, you are asking about whether we are able to use BVH algorithms implemented by others or load precomputed BVH into this renderer. The answer is yes but it comes at a cost: we need to implement the correct interface, which converts the BVH outputs from whatever code it is, to Taichi lang `ti.dataclass`. This is the only modification my renderer requires, that is, you just need to correctly feed the constructed BVH and everything is good to go.

However, note that this BVH implementation is **suboptimal**. Here is the situation: Taichi lang has limited ability to dynamically allocate memory inside of its kernel function, but BVH tree traversal (the optimal one) requires building a stack (otherwise, use recursive calls). I didn't find a easy way to build a stack in the kernel function (actually, `dynamic` node in Taichi lang might help but I'm not sure whether it can be a fast implementation), so the tree traversal is done **linearly**. There is going-backwards: if the `lchild` is accessed first and pruned, I can directly jump to the `rchild` and this would accelerate the rendering since we do skip some impossible ray-intersections, but I can't do this the other way around, i.e I can't check `rchild` first then jump back to `lchild`. The implementation could get messy so I stuck to the easier and more intuitive way.

**Reply by kurt-rhee** [reply]
Wow, very impressed.  I wasn't able to find the BVH in your code, would you be able to point me to it?  My understanding of the PBR book based off of their blog is that there are multiple ways of constructing a BVH, (top down, bottoms up) I am interested in which one you picked.

**Reply by Enigmatisms** [reply, marked as answer]
BVH implementation can be found here: https://github.com/Enigmatisms/AdaPT/tree/master/tracer/bvh. This is implemented in C++ and I use pybind11 to export an interface for python.

I picked a top-down approach with SAH function, which should be the one introduced in [PBR book 4.3.2](https://pbr-book.org/3ed-2018/Primitives_and_Intersection_Acceleration/Bounding_Volume_Hierarchies#TheSurfaceAreaHeuristic)
